refactor(metrics): move from protocol to ABC; split provider tests

leandrodamascena opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** #2916 

## Summary

### Changes

This PR addresses the latest issues before we add the Datadog Metrics provider. There are no changes for the customer or even experience, this just fixes internal things.

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

leandrodamascena committed:
Removing old code

leandrodamascena committed:
Making function standalone

leandrodamascena committed:
Fixing docstring

leandrodamascena committed:
Removing subclassing + adding unit tests

heitorlessa commented on aws_lambda_powertools/metrics/provider/base.py:
```diff
@@ -4,16 +4,17 @@
 
 
+@runtime_checkable
 class MetricsProviderBase(Protocol):
```
Let's convert all this as an ABC which requires certain methods, has docstrings, and implements Powertools for AWS Lambda value add (e.g., log_metrics, etc.).

It looked like a good idea to have Protocol when we were to receive a `MetricsProvider` within `Metrics`, but that failed due to the number of specializations for each provider.

We will also not need this anymore, merge with `MetricsBase` and actually call them `BaseProvider(ABC)`.

heitorlessa commented on aws_lambda_powertools/metrics/provider/base.py:
```diff
@@ -25,80 +26,19 @@ class MetricsProviderBase(Protocol):
+        ...
 
 
+@runtime_checkable
```
```suggestion
```

heitorlessa commented on aws_lambda_powertools/metrics/provider/base.py:
```diff
@@ -25,80 +26,19 @@ class MetricsProviderBase(Protocol):
 
 
+@runtime_checkable
 class MetricsBase(Protocol):
```
```suggestion
class BaseProvider(ABC):
```

heitorlessa commented on tests/unit/metrics/test_metrics.py:
make it specific to CloudWatch tests

`test_cloudwatch_emf_provider.py`

This allows us to have specific tests for the Provider capability, and only use `log_metrics` in a functional tests since multiple units will be covered.

heitorlessa replied:
This also allows us to clean up our tests within `test_metrics` and have all specialized CW tests for the provider itself.

and faster tests since unit tests will be small in nature.

heitorlessa commented on tests/unit/metrics/test_metrics.py:
```diff
@@ -0,0 +1,8 @@
+from aws_lambda_powertools.metrics.provider.cloudwatch_emf.cloudwatch import AmazonCloudWatchEMFProvider
+
+
+def test_amazoncloudwatchemf_is_subclass_of_metricsproviderbase():
+    assert issubclass(AmazonCloudWatchEMFProvider, MetricsProviderBase)
```
we're not gonna need it after we go with ABC

heitorlessa reviewed (changes requested):
peer review comments

leandrodamascena committed:
Refactoring metrics

leandrodamascena committed:
Refactoring tests

leandrodamascena committed:
Moving files

leandrodamascena committed:
Cleaning code

heitorlessa commented on aws_lambda_powertools/metrics/provider/base.py:
```diff
@@ -2,19 +2,17 @@
+    Class for metric provider interface.
 
     This class serves as an interface for creating your own metric provider. Inherit from this class
     and implement the required methods to define your specific metric provider.
```
```suggestion
    Interface to create a metrics provider.

    BaseProvider implements `log_metrics` decorator for every provider as a value add feature.
```

We can keep it short here since there was duplicate. Inform whoever is implementing it that `log_metrics` is a value add unless they want to override.

If the value add features grow, we can make into a list later.

heitorlessa commented on aws_lambda_powertools/metrics/provider/base.py:
```diff
@@ -25,41 +23,127 @@ class MetricsProviderBase(Protocol):
-    def add_metric(self, *args, **kwargs):
-        ...
+    @abstractmethod
+    def add_cold_start_metric(self, context: Any) -> Any:
```
```suggestion
    def add_cold_start_metric(self, context: LambdaContext) -> Any:
```

heitorlessa commented on aws_lambda_powertools/metrics/provider/base.py:
```diff
@@ -25,41 +23,127 @@ class MetricsProviderBase(Protocol):
+        """
+        Abstract method for clear metric instance.
+
+        This method must be implemented in subclasses to add a metric and return a combined metrics dictionary.
```
I think you copied from `add_metric`

heitorlessa commented on aws_lambda_powertools/metrics/provider/cloudwatch_emf/cloudwatch.py:
```diff
@@ -12,30 +11,31 @@
-    """Base class for metric functionality (namespace, metric, dimension, serialization)
+class AmazonCloudWatchEMFProvider(BaseProvider):
+    """
+    AmazonCloudWatchEMFProvider class (namespace, metric, dimension, serialization)
```
```suggestion
    AmazonCloudWatchEMFProvider creates metrics asynchronously via CloudWatch Embedded Metric Format (EMF).
```

heitorlessa commented on aws_lambda_powertools/metrics/provider/base.py:
```diff
@@ -104,11 +190,14 @@ def handler(event, context):
         def decorate(event, context):
             try:
+                if default_dimensions:
+                    self.set_default_dimensions(**default_dimensions)
```
this method doesn't exist in Base. What you'd want to do is keep everything but the logic for `default_dimensions`.

Then override `log_metrics` in `AmazonCloudWatchEMFProvider` to set the default dimensions if available, then call `super().log_metrics(...)`

leandrodamascena replied:
The reason I kept it in the base class is because we are subclassing, so we access the `default_dimensions` of the child class. But yes, it makes a lot of sense to keep provider specializations in their specific class.

Fixed.

heitorlessa commented on aws_lambda_powertools/metrics/provider/base.py:
```diff
@@ -120,24 +209,23 @@ def decorate(event, context):
         return decorate
 
     def _add_cold_start_metric(self, context: Any) -> None:
```
```suggestion
    def _add_cold_start_metric(self, context: LambdaContext) -> None:
```

heitorlessa commented on aws_lambda_powertools/metrics/provider/base.py:
```diff
@@ -120,24 +209,23 @@ def decorate(event, context):
     def _add_cold_start_metric(self, context: Any) -> None:
-        """Add cold start metric and function_name dimension
+        """
+        Check if it's cold start and add a metric if yes
```
```suggestion
        Add cold start metric
```

heitorlessa commented on tests/functional/metrics/test_metrics_provider.py:
```diff
@@ -0,0 +1,182 @@
+
+
+@pytest.fixture
+def metrics_provider() -> BaseProvider:
```
is that correct? looks like a leftover from the Protocol implementation. Now that we have an ABC, it'd fail without `clear_metrics`, but it doesn't because we're implementing a standalone class.

heitorlessa commented on tests/functional/metrics/test_metrics_provider.py:
```diff
@@ -0,0 +1,182 @@
+    provider.flush()
+    output = capture_metrics_output(capsys)
+    assert output[0]["name"] == metric["name"]
+    assert output[0]["value"] == metric["value"]
```
comment above on whether this is valid

heitorlessa reviewed (changes requested):
Almost there, I promise! `log_metrics` is incorrect, and I think there's some leftover from the Protocol implementation within tests.

leandrodamascena committed:
Merge branch 'develop' into techdebt/metrics

leandrodamascena committed:
Addressing Heitor's feedback

heitorlessa commented:
Quick patch to type CloudWatch EMF Output, it's an old old tech debt I always wanted to pay ever since we added typing extensions -- this will make asserting on tests easier too (and for customers!).

[0001-refactor-type-EMF-output-with-TypedDict.patch](https://github.com/aws-powertools/powertools-lambda-python/files/12300989/0001-refactor-type-EMF-output-with-TypedDict.patch)

leandrodamascena committed:
Addressing Heitor's feedback

leandrodamascena commented:
> Quick patch to type CloudWatch EMF Output, it's an old old tech debt I always wanted to pay ever since we added typing extensions -- this will make asserting on tests easier too (and for customers!).
> 
> [0001-refactor-type-EMF-output-with-TypedDict.patch](https://github.com/aws-powertools/powertools-lambda-python/files/12300989/0001-refactor-type-EMF-output-with-TypedDict.patch)

I applied this patch and addressed all feedback. Can you review it, please?

heitorlessa commented:
checking 👀 !

heitorlessa commented on aws_lambda_powertools/metrics/provider/cloudwatch_emf/cloudwatch.py:
```diff
@@ -18,16 +18,18 @@
-    AmazonCloudWatchEMFProvider class (namespace, metric, dimension, serialization)
+    AmazonCloudWatchEMFProvider creates metrics asynchronously via CloudWatch Embedded Metric Format (EMF).
 
     AmazonCloudWatchEMFProvider creates metrics asynchronously thanks to CloudWatch Embedded Metric Format (EMF).
```
duplicated?

heitorlessa commented on tests/functional/metrics/test_metrics_provider.py:
```diff
@@ -1,101 +1,113 @@
-            self.result = ""
-            self.metric_store = []
+    def clear_metrics(self):
+        self.metric_store = []
```
```suggestion
        self.metric_store.clear()
```

intent

heitorlessa commented on tests/functional/metrics/test_metrics_provider.py:
```diff
@@ -1,101 +1,113 @@
+            capture_cold_start_metric=capture_cold_start_metric,
+            raise_on_empty_metrics=raise_on_empty_metrics,
+        )
 
```
you don't need it since it's in the BaseProvider (the beauty of it) :)

heitorlessa commented on tests/functional/metrics/test_metrics_provider.py:
```diff
@@ -1,101 +1,113 @@
+            capture_cold_start_metric=capture_cold_start_metric,
+            raise_on_empty_metrics=raise_on_empty_metrics,
+        )
 
```
we don't need it anymore ;)

heitorlessa commented on tests/functional/metrics/test_metrics_provider.py:
```diff
@@ -1,101 +1,113 @@
-    provider.add_metric(**metric)
-    provider.serialize_metric_set()
-    provider.flush()
+def test_metrics_class_with_default_provider(capsys, metric):
```
we don't need it anymore as we have this tested under functional metrics `Metrics()` ;)

heitorlessa commented on tests/functional/metrics/test_metrics_provider.py:
```diff
@@ -1,101 +1,113 @@
-    metrics = metrics_class(provider=metrics_provider())
+def test_metrics_class_with_custom_provider(capsys, metric):
+    provider = FakeMetricsProvider()
+    metrics = FakeMetricsClass(provider=provider)
```
metrics = Metrics(provider=provider)

So we adhere to test the actual provider not test a Metrics class since we already have it ;)

> repeat in other areas

heitorlessa commented on tests/functional/metrics/test_metrics_provider.py:
```diff
@@ -1,101 +1,113 @@
+            lambda_handler=lambda_handler,
+            capture_cold_start_metric=capture_cold_start_metric,
+            raise_on_empty_metrics=raise_on_empty_metrics,
+        )
```
we don't need anymore since we're testing the custom provider ability, our `Metrics` should do it ;)

heitorlessa reviewed (changes requested):
last changes as we luckily don't need those extra bits in testing ;)

leandrodamascena committed:
Addressing Heitor's feedback
